overlord/snapstate/snapstate.go

```diff
@@ -273,6 +274,89 @@ func isCoreSnap(snapName string) bool {
 	return snapName == defaultCoreSnapName
 }
 
+// removeExtraComponentsTasks creates tasks that will remove unwanted components
+// that are currently installed alongside the snap revision that is about to be
+// installed. If the new snap revision is not in the sequence, then we don't
+// have anything to do. If the revision is in the sequence, then we generate
+// tasks that will unlink components that are not in compsups.
+//
+// This is mostly relevant when we're moving from one snap revision to another
+// snap revision that has already been installed on the system. The target snap
+// might have components that are installed that we don't want any more.
```

Comment:
I'm confused by this comment, aren't the component linked to a given snap revision, why would we remove components when switch around by revision?

Reply:
Yes, they are linked by snap revision. Consider this case:
```
snap revision 1, components a, b
snap revision 2, components b, c
```

When we move from snap revision 2 -> 1, we need to unlink component a (and discard it, if nothing else references it) and install component c.

Reply:
I am a bit confused about the comment too, should "target snap might have components that are installed that we don't want any more" be "target revision might have components that were installed in the past but that we don't want any more"? Or maybe state more clearly that this is about checking a revision that was active in the past and still around but not currently. I always find misleading when we talk about installed revisions for non-active but present revisions.

Reply:
I see what you mean, I'll update it to be explicit when I'm talking about something that was installed in the past.